x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/plugin/SqlMediaTypeParser.java

```diff
@@ -0,0 +1,77 @@
+import org.elasticsearch.rest.RestRequest;
+import org.elasticsearch.xpack.sql.action.SqlQueryRequest;
+import org.elasticsearch.xpack.sql.proto.Mode;
+
+import java.util.Map;
+import java.util.regex.Pattern;
+
+import static org.elasticsearch.xpack.sql.proto.Protocol.URL_PARAM_FORMAT;
+
+public class SqlMediaTypeParser {
+    private static final MediaTypeParser<? extends MediaType> parser = new MediaTypeParser.Builder<>()
+        .copyFromMediaTypeParser(XContentType.mediaTypeParser)
+        .withMediaTypeAndParams(TextFormat.PLAIN_TEXT.typeWithSubtype(), TextFormat.PLAIN_TEXT,
+            Map.of("header", Pattern.compile("present|absent"), "charset", Pattern.compile("utf-8")))
+        .withMediaTypeAndParams(TextFormat.CSV.typeWithSubtype(), TextFormat.CSV,
+            Map.of("header", Pattern.compile("present|absent"), "charset", Pattern.compile("utf-8")))
+        .withMediaTypeAndParams(TextFormat.TSV.typeWithSubtype(), TextFormat.TSV,
+            Map.of("header", Pattern.compile("present|absent"), "charset", Pattern.compile("utf-8")))
+        .build();
+
+    /*
+     * Since we support {@link TextFormat} <strong>and</strong>
+     * {@link XContent} outputs we can't use {@link RestToXContentListener}
+     * like everything else. We want to stick as closely as possible to
+     * Elasticsearch's defaults though, while still layering in ways to
+     * control the output more easily.
+     *
+     * First we find the string that the user used to specify the response
+     * format. If there is a {@code format} parameter we use that. If there
+     * isn't but there is a {@code Accept} header then we use that. If there
+     * isn't then we use the {@code Content-Type} header which is required.
+     */
+    public MediaType getMediaType(RestRequest request, SqlQueryRequest sqlRequest) {
+
+        if (Mode.isDedicatedClient(sqlRequest.requestInfo().mode())
+            && (sqlRequest.binaryCommunication() == null || sqlRequest.binaryCommunication())) {
+            // enforce CBOR response for drivers and CLI (unless instructed differently through the config param)
+            return XContentType.CBOR;
+        } else if (request.hasParam(URL_PARAM_FORMAT)){
+            return validateColumnarRequest(sqlRequest.columnar(), parser.fromFormat( request.param(URL_PARAM_FORMAT)));
+        }
+        if (request.getHeaders().containsKey("Accept")) {
+            String accept = request.header("Accept");
+            // */* means "I don't care" which we should treat like not specifying the header
+            if ("*/*".equals(accept) == false) {
+                return validateColumnarRequest(sqlRequest.columnar(), parser.fromMediaType(accept));
+            }
+        }
+
+        //todo pg we should probably remove this possibility. it should only be accept or format.
+        // also if you specify sql's textformat in content type it will fail with exception now
+        String contentType = request.header("Content-Type");
```

Comment:
I made a comment to suggest we get rid of using Content-Type to set responseType. However I think most sql queries contain a body, so using a Content-Type to default responseType might be ok.
If however a request did not contain a body, then it would not make sense. I don't know we have any SQL api though.
However, this obviously means that only json/cbor/smile/yml are allowed .

Reply:
also not fixing this in current PR. might be worth discussing with @elastic/es-ql

Reply:
> I think most sql queries contain a body

Indeed, all requests concerning this code always contain a body.
(In the SQL API only one API lacks it, the one for stats inquiring, which won't hit this code tho.)